pyproject.toml

```diff
@@ -0,0 +1,178 @@
+  "Programming Language :: Python :: 3.7", # TODO: EOL 2023-06-27
+  "Programming Language :: Python :: 3.8",
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",
+]
+dynamic = [
+  "version",
+]
+dependencies = [
+  "defusedcsv~=2.0.0",
+  "defusedxml~=0.7.1",
+  "Django~=3.2.20",
+  "django-allauth~=0.51.0",
+  "django-cleanup~=6.0.0",
+  "django-compressor~=4.1", # 4.2 drops Python 3.6 and 3.7 support, https://django-compressor.readthedocs.io/en/stable/changelog.html#v4-2-2023-01-06
+  "django-extensions~=3.2.0",
+  "django-filter~=21.1",
+  "django-libsass==0.9",
+  "django-mathfilters~=1.0.0",
+  "django-mptt~=0.13.4",
+  "django-rest-swagger~=2.2.0",
+  "django-settings-export~=1.2.1",
+  "django-widget-tweaks~=1.4.8",
+  "djangorestframework~=3.12.4",
+  "drf-extensions~=0.7.0",
+  "iso8601~=2.0.0", # 2.0.0 drops Python 3.6 support, https://github.com/micktwomey/pyiso8601/blob/main/CHANGELOG.md
+  "jsonfield~=3.1.0",
+  "Markdown~=3.3.7",
+  "pypandoc~=1.10.0",
+  "rules~=3.3.0",
+]
+[project.optional-dependencies]
+ci = [
+  "coveralls",
+  "rdmo[dev]",
+]
+dev = [
+  "django-debug-toolbar",
+  "pre-commit",
+  "rdmo[tests]",
+]
+docs = [
+  "recommonmark",
+  "sphinx",
+  "sphinx-autobuild",
+  "sphinx_rtd_theme",
+]
+gunicorn = [
+  "gunicorn>=19.9",
+]
+ldap = [
+  "django-auth-ldap>=2",
+]
+memcached = [
+  "python-memcached>=1.58",
+]
+mysql = [
+  "mysqlclient~=2.1.0", # 2.2.0 drops Python 3.7 support, https://github.com/PyMySQL/mysqlclient/blob/main/HISTORY.rst#whats-new-in-220
+]
+postgres = [
+  "psycopg2-binary~=2.9.6",
+]
+shibboleth = [
+  "django-shibboleth-remoteuser@ git+https://github.com/Brown-University-Library/django-shibboleth-remoteuser.git@main",
+]
+tests = [
+  "pytest~=7.1.0", # 7.1.0 drops Python 3.6 support, https://docs.pytest.org/en/stable/changelog.html#pytest-7-1-0-2022-03-13
+  "pytest-cov~=4.0.0",
+  "pytest-django~=4.5.0",
+  "pytest-dotenv~=0.5.2",
+  "pytest-mock~=3.7.0", # 3.7.0 drops Python 3.6 support, https://pytest-mock.readthedocs.io/en/latest/changelog.html
+  "pytest-randomly~=3.12.0", # 3.13.0 drops Python 3.7 support, https://github.com/pytest-dev/pytest-randomly/blob/main/CHANGELOG.rst#3120-2022-05-11
+  "pytest-xdist~=3.3.1",
+]
+[project.urls]
+Changelog = "https://github.com/rdmorganiser/rdmo/blob/master/CHANGELOG.md"
+Documentation = "https://rdmo.readthedocs.io"
+Homepage = "https://rdmorganiser.github.io/"
+Issues = "https://github.com/rdmorganiser/rdmo/issues"
+Repository = "https://github.com/rdmorganiser/rdmo.git"
+
+[tool.setuptools]
+packages = ["rdmo"]
+
+[tool.setuptools.dynamic]
+version = {attr = "rdmo.__version__"}
+
+[tool.black]
+line-length = 120
+skip-string-normalization = true
+target_version = ["py37", "py38", "py39", "py310", "py311"] # TODO: EOL 2023-06-27
+
+[tool.ruff]
+target-version = "py37" # TODO: EOL 2023-06-27
+line-length = 120
+select = [
```

Comment:
I tired the rules one by one and I would add (or better remove):

```
ignore = [
  "B006",   # mutable data structures for argument defaults
  "B007",   # loop control variable not used within loop body
  "RUF012", # mutable class attributes should be annotated with `typing.ClassVar`
  "UP031",  # use format specifiers instead of percent format
  "UP032",  # use f-string instead of `format` call
]
```

I don't consider the tow B rules so bad, it can improve readbility if multiple loops look the same. I am not a fan of types, but not against them, but I don't want to be forced to use them. The strings are just not so important. I try to use f strings now, but I don't want to refactor all the % strings.

Reply:
I took the ruff ruleset from [build](https://github.com/pypa/build/blob/main/pyproject.toml#L142C7-L142C7) as discussed today, without mccabe.

Ruff can actually "fix" a lot of these issues automatically, like UP031 and UP032.

The diff after reformatting would look like this:
https://github.com/afuetterer/rdmo/pull/5/files

Still ruff complains about some rules, like RUF012. But yeah, I can ignore them.

What do you think about the diff?

Reply:
Yes UP031 and UP032 make sense and the fix looks good. I would only add the three other ignores. All the remaining errors should be fixed either automatically or manually. I was really impressed with ruff btw.

Reply:
Alright, I will add those ignores.
Nice, that you enjoy ruff as well. I like it a lot.